components/sup/src/ctl_gateway/server.rs

```diff
@@ -0,0 +1,428 @@
+    // argument passed to this closure until `FnOnce<Box>` stabilizes.
+    //
+    // https://github.com/rust-lang/rust/issues/28796
+    fun: Box<Fn(&ManagerState, &mut CtlRequest) -> NetResult<()> + Send>,
+}
+
+impl CtlCommand {
+    /// Create a new CtlCommand from the given CtlSender, transaction, and closure to execute.
+    pub fn new<F>(tx: Option<CtlSender>, txn: Option<SrvTxn>, fun: F) -> Self
+    where
+        F: Fn(&ManagerState, &mut CtlRequest) -> NetResult<()> + Send + 'static,
+    {
+        CtlCommand {
+            fun: Box::new(fun),
+            req: CtlRequest::new(tx, txn),
+        }
+    }
+
+    /// Run the contained closure with the given [`manager.ManagerState`].
+    pub fn run(&mut self, state: &ManagerState) -> NetResult<()> {
+        (self.fun)(state, &mut self.req)
+    }
+}
+
+/// Server's client representation. Each new connection will allocate a new Client.
+struct Client {
+    handle: reactor::Handle,
+    state: Rc<RefCell<SrvState>>,
+}
+
+impl Client {
+    /// Serve the client from the given framed socket stream.
+    pub fn serve(self, socket: SrvStream) -> Box<Future<Item = (), Error = HandlerError>> {
+        let mgr_tx = self.state.borrow().mgr_tx.clone();
+        Box::new(self.handshake(socket).and_then(|socket| {
+            SrvHandler::new(socket, mgr_tx)
+        }))
+    }
+
+    /// Initiate a handshake with the connected client before allowing future requests. A failed
+    /// handshake will close the connection.
+    fn handshake(&self, socket: SrvStream) -> Box<Future<Item = SrvStream, Error = HandlerError>> {
+        let secret_key = self.state.borrow().secret_key.to_string();
+        let handshake = socket
+            .into_future()
+            .map_err(|(err, _)| HandlerError::from(err))
+            .and_then(move |(m, io)| {
+                m.map_or_else(
+                    || {
+                        Err(HandlerError::from(
+                            io::Error::from(io::ErrorKind::UnexpectedEof),
+                        ))
+                    },
+                    move |m| {
+                        if m.message_id() != "Handshake" {
+                            debug!("No handshake");
+                            return Err(HandlerError::from(
+                                io::Error::from(io::ErrorKind::ConnectionAborted),
+                            ));
+                        }
+                        if !m.is_transaction() {
+                            return Err(HandlerError::from(
+                                io::Error::from(io::ErrorKind::ConnectionAborted),
+                            ));
+                        }
+                        match m.parse::<protocols::ctl::Handshake>() {
+                            Ok(decoded) => {
+                                trace!("Received handshake, {:?}", decoded);
+                                Ok((m, decoded.get_secret_key() == secret_key, io))
```

Comment:
It occurs to me that we probably want to use a constant-time comparison here (such as [`crypto::util::fixed_time_eq`](https://docs.rs/rust-crypto/0.2.36/crypto/util/fn.fixed_time_eq.html)) or we may be opening ourselves up to timing attacks.